### PR TITLE
Ecosystem page: initial revision

### PR DESCRIPTION
**Pull request by fanciulli**
Very first commit. Opening this PR is meant to have an initial contact and start discussion.

What has been done:

1. Added Markdown page for Ecosystem under src/pages/ecosystem.mdx
2. Create Two react components to handle: 1) Summary line with plugin  count, 2) Plugin table
3. Created JSON file with plugins data which is read by the Javascript file initializing the React components

What has to be done next:

1. Descriptions still include the original Markdown sections. To understand if it is possible to have Docasaurus to parse it (probably not) or have a library translate to links
2. Replace current table with a React table with filtering etc...
3. Maybe add some styling (event though this could be a website wide task, target for another ticket)

Any suggestion is welcome as i am not a great expert in React (learning right now).

NOTE: i had to comment the reference to the index doc in docasaurus.config.js as it was preventing from properly running. Same for the versions.json file as those seems to require content in the versioned_docs and versioned_sidebars forlders

### CODE REVIEW

**Review by Eomm — COMMENTED** [review]
> Maybe add some styling (event though this could be a website wide task, target for another ticket)

Yeah, totally agree with that (for example the `input` box).
Regardless the style, I think we need to write some CSS lines to manage the plugin grid and a bit the width:

image


> Any suggestion is welcome as i am not a great expert in React (learning right now).

I add some comments and discuss them - don't feel pressure on it 😄

> NOTE: i had to comment the reference to the index doc in docasaurus.config.js as it was preventing from properly running. Same for the versions.json file as those seems to require content in the versioned_docs and versioned_sidebars folders

You did well, I think we will get a stable website next month

Reminders:

- I think there is a way to render the markdown description but we will need some `eval` style command for sure
- we need an empty state for plugins (no results)

Well done so far 🎉🚀

File: src/pages/ecosystem.mdx
Comment (fanciulli) [reply src/pages/ecosystem.mdx:14]: Adding in a single commit

File: src/pages/ecosystem.mdx
Comment (fanciulli) [reply src/pages/ecosystem.mdx:20]: Adding in a single commit

File: src/components/ecosystem/plugins.js
Comment (fanciulli) [reply src/components/ecosystem/plugins.js:6]: Went this way because i needed to implement a callback function to call with onKeyUp.

Was there another way to do this?

File: src/components/ecosystem/plugins.js
Comment (fanciulli) [reply src/components/ecosystem/plugins.js:1]: Docusaurus and examples in the project all use the .js extension. Will update anyway as we are putting React components there..

File: src/components/ecosystem/plugins.js
Comment (fanciulli) [reply src/components/ecosystem/plugins.js:73]: Sure.

File: src/pages/ecosystem.mdx
Comment (fanciulli) [reply src/pages/ecosystem.mdx:10]: OK

File: src/pages/ecosystem.mdx
Comment (fanciulli) [reply src/pages/ecosystem.mdx:7]: Absolutely. Will implement.

File: src/components/ecosystem/plugins.js
Comment (fanciulli) [reply src/components/ecosystem/plugins.js:49]: I skipped using react table also because i neither wanted to add plugins for just a simple table.
Going to table a look at the CSS grids and implement that

File: src/components/ecosystem/plugins.json
Comment (fanciulli) [reply src/components/ecosystem/plugins.json:2]: I started from your suggestion and created a script (which i haven't committed). Takes the Ecosystem.md as input and writes the output in console.log

Will add the script. Where exactly? is the scripts folder ok?

File: versions.json
Comment (HassanBahati) [inline versions.json:1]: @fanciulli was this your solution to this error?
`[ERROR] Error: The docs folder does not exist for version "v3.29.4". A docs folder is expected to be found at versioned_docs/version-v3.29.4.
`

Comment (HassanBahati) [inline versions.json:1]: @fanciulli @Eomm i have opened a [pr](https://github.com/fastify/website-next/pull/37) #37  to address this based on the fastify current versions...hopefully that would solve the issue you mentioned in the in the pr description about an error that was popping up.

Comment (fanciulli) [reply versions.json:1]: @HassanBahati very nice, thank you. I will merge it as soon as PR is accepted

File: src/components/ecosystem/plugins.js
Comment (fanciulli) [reply src/components/ecosystem/plugins.js]: Done

**Comment by fanciulli** [comment]
Hi @Eomm @HassanBahati 

the PR is complete. I have implemented everything that was reported/suggested. Waiting for approval or other comments.

File: versions.json
Comment (fanciulli) [reply versions.json:1]: Not doing anything since it seems to me that PR is changing the file to same exact value

**Review by Eomm — COMMENTED** [review]
Well done!
We are near the finish line. I left some suggestions, please don't curse me 😄

---

Looking at the console I see this error:

image

There are some `key` attributes that must be set on the components - I did not have the time to spot them, but we need solve this error. Let me know if you need more help!

File: package.json
Comment (fanciulli) [reply package.json:25]: Yep. removed react-router-dom.
indeed i was using <Link> from docusaurus so this one is a leftover of a previoous implementation (which was using react-router)

File: src/pages/ecosystem.mdx
Comment (fanciulli) [reply src/pages/ecosystem.mdx:3]: You mean importing the JSON file here?
It is a better approach for sure

File: src/components/ecosystem/plugins_table.jsx
Comment (fanciulli) [reply src/components/ecosystem/plugins_table.jsx]: Done

File: src/components/ecosystem/plugins_table.jsx
Comment (fanciulli) [reply src/components/ecosystem/plugins_table.jsx]: ok, no problem

**Comment by fanciulli** [comment]
@Eomm Implemented all suggestions.

**Comment by HassanBahati** [comment]
@fanciulli well done :clap: thank you for the great work on here! 

I have a suggestion though; since these plugins are many, can we have a sidebar to the left that indexes them, I feel that navigation is quite hard without a means of indexing...since we are having an empty white space there.
Screenshot from 2023-01-01 21-00-37

**Comment by fanciulli** [comment]
@HassanBahati grouping via a left panel can help since plugins are many (250+). 
In my experience people are not always searching things knowing the group these are. This means that we also need to think more about the filtering capability to work better with this change.

I think that this should go through a separate task, starting with a "requirements" collection

**Review by Eomm — APPROVED** [review]
> what do you think ?

I think the entire website requires a UX/UI overview. We talk about adding `topics` into the GH repositories to add some tags in this layout. We could merge the two plugin tables and add a column that say "official" or "unofficial" - or even the latest fastify version supported (I'm sure some of those 200 plugins are not updated to fastify v4)

There is a lot to do/ideas but the first step is always creating a new issue to define the target to reach - so feel free to open it and list all your idea 👍🏽 

--- 

This PR has hit its target and LGTM!
Thank you very much @fanciulli

**Pull request merged**
